Rest server

cschuchardt88 opened the pull request:
## RestServer
In this section you will learn about `RestServer` plugin and how it works.

## Dependencies
- **Microsoft.AspNetCore.JsonPatch.dll** `Required`
- **Microsoft.AspNetCore.Mvc.NewtonsoftJson.dll** `Required`
- **Microsoft.AspNetCore.Mvc.Versioning.ApiExplorer.dll** `Required`
- **Microsoft.AspNetCore.Mvc.Versioning.dll** `Required`
- **Microsoft.OpenApi.dll** `Required`
- **Newtonsoft.Json.Bson.dll** `Required`
- **Newtonsoft.Json.dll** `Required`
- **System.ServiceProcess.ServiceController.dll** `linux maybe`
- **Microsoft.AspNetCore.Mvc.Versioning.dll** `Required`
- **Microsoft.AspNetCore.Mvc.Versioning.dll** `Required`
- **Microsoft.AspNetCore.Mvc.Versioning.dll** `Required`
- **Microsoft.OpenApi.dll** `Swagger` _(optional)_
- **Swashbuckle.AspNetCore.Swagger.dll** `Swagger` _(optional)_
- **Swashbuckle.AspNetCore.SwaggerGen.dll** `Swagger` _(optional)_
- **Swashbuckle.AspNetCore.SwaggerUI.dll** `Swagger` _(optional)_
- **Swashbuckle.AspNetCore.Newtonsoft.dll** `Swagger` _(optional)_
- **RestServer.xml** `Swagger UI` _(optional)_

### In Docker
```
RestServer
|   |   |-- Microsoft.AspNetCore.JsonPatch.dll
|   |   |-- Microsoft.AspNetCore.Mvc.NewtonsoftJson.dll
|   |   |-- Microsoft.AspNetCore.Mvc.Versioning.ApiExplorer.dll
|   |   |-- Microsoft.AspNetCore.Mvc.Versioning.dll
|   |   |-- Microsoft.OpenApi.dll
|   |   |-- Neo.ConsoleService.dll
|   |   |-- Newtonsoft.Json.Bson.dll
|   |   |-- Newtonsoft.Json.dll
|   |   |-- RestServer.deps.json
|   |   |-- RestServer.dll
|   |   |-- RestServer.pdb
|   |   |-- RestServer.xml
|   |   |-- Swashbuckle.AspNetCore.Newtonsoft.dll
|   |   |-- Swashbuckle.AspNetCore.Swagger.dll
|   |   |-- Swashbuckle.AspNetCore.SwaggerGen.dll
|   |   |-- Swashbuckle.AspNetCore.SwaggerUI.dll
|   |   |-- System.ServiceProcess.ServiceController.dll
|   |   |-- config.json
|   |   `-- runtimes
|   |       `-- win
|   |           `-- lib
|   |               `-- net6.0
|   |                   `-- System.ServiceProcess.ServiceController.dll
```

---

These files go in the same directory as the `RestServer.dll`. In neo-cli
`plugins/RestServer/` folder.

## Response Headers
| Name | Value(s) | Description |
| :---: | --- | :--- |
|**server**|_neo-cli/3.6.0 RestServer/3.6.0_|_`neo-cli` and `RestServer` version._|

## JSON Serializer
`RestServer` uses custom Newtonsoft Json Converters to serialize controller action
responses and `route` parameters.

**One Way Binding** - `Write` only.
- `Neo.SmartContract.ContractState`
- `Neo.SmartContract.NefFile`
- `Neo.SmartContract.MethodToken`
- `Neo.SmartContract.Native.TrimmedBlock`
- `Neo.SmartContract.Manifest.ContractAbi`
- `Neo.SmartContract.Manifest.ContractGroup`
- `Neo.SmartContract.Manifest.ContractManifest`
- `Neo.SmartContract.Manifest.ContractPermission`
- `Neo.SmartContract.Manifest.ContractPermissionDescriptor`
- `Neo.Network.P2P.Payloads.Block`
- `Neo.Network.P2P.Payloads.Header`
- `Neo.Network.P2P.Payloads.Signer`
- `Neo.Network.P2P.Payloads.TransactionAttribute`
- `Neo.Network.P2P.Payloads.Transaction`
- `Neo.Network.P2P.Payloads.Witness`

**Two Way Binding** - `Read` & `Write`
- `System.Guid`
- `System.ReadOnlyMemory<T>`
- `Neo.BigDecimal`
- `Neo.UInt160`
- `Neo.UInt256`
- `Neo.Cryptography.ECC.ECPoint`
- `Neo.VM.Types.Array`
- `Neo.VM.Types.Boolean`
- `Neo.VM.Types.Buffer`
- `Neo.VM.Types.ByteString`
- `Neo.VM.Types.Integer`
- `Neo.VM.Types.InteropInterface`
- `Neo.VM.Types.Null`
- `Neo.VM.Types.Map`
- `Neo.VM.Types.Pointer`
- `Neo.VM.Types.StackItem`
- `Neo.VM.Types.Struct`

## Remote Endpoints
Parametes `{hash}` can be any Neo N3 address or scripthash; `{address}` can be any Neo N3 address **only**; `{number}` and `{index}` can be any _**uint32**_.

**Parameter Examples**
- `{hash}` - _0xef4073a0f2b305a38ec4050e4d3d28bc40ea63f5_ **or** _NiHURyS83nX2mpxtA7xq84cGxVbHojj5Wc_
- `{address}` - _NiHURyS83nX2mpxtA7xq84cGxVbHojj5Wc_
- `{number}` - _1_
- `{index}` - _2500000_

**Paths**
- Utils
  - `[GET]` `/api/v1/utils/{hash}/address`
  - `[GET]` `/api/v1/utils/{address}/scripthash`
  - `[GET]` `/api/v1/utils/{hash}/{address}/validate`
- Node
  - `[GET]` `/api/v1/node`
  - `[GET]` `/api/v1/node/peers`
  - `[GET]` `/api/v1/node/plugins`
  - `[GET]` `/api/v1/node/settings`
- Ledger
  - `[GET]` `/api/v1/ledger/neo/accounts`
  - `[GET]` `/api/v1/ledger/gas/accounts`
  - `[GET]` `/api/v1/ledger/blocks?page={number}&size={number}`
  - `[GET]` `/api/v1/ledger/blocks/height`
  - `[GET]` `/api/v1/ledger/blocks/{index}`
  - `[GET]` `/api/v1/ledger/blocks/{index}/header`
  - `[GET]` `/api/v1/ledger/blocks/{index}/witness`
  - `[GET]` `/api/v1/ledger/blocks/{index}/transactions?page={number}&size={number}`
  - `[GET]` `/api/v1/ledger/transactions/{hash}`
  - `[GET]` `/api/v1/ledger/transactions/{hash}/witnesses`
  - `[GET]` `/api/v1/ledger/transactions/{hash}/signers`
  - `[GET]` `/api/v1/ledger/transactions/{hash}/attributes`
  - `[GET]` `/api/v1/ledger/memorypool?page={number}&size={number}`
  - `[GET]` `/api/v1/ledger/memorypool/verified?page={number}&size={number}`
  - `[GET]` `/api/v1/ledger/memorypool/unverified?page={number}&size={number}`
  - `[GET]` `/api/v1/ledger/memorypool/count`
- Tokens
  - `[GET]` `/api/v1/tokens/balanceof/{address}`
  - NFTs
    - `[GET]` `/api/v1/tokens/nep-11?page={number}&size={number}`
    - `[GET]` `/api/v1/tokens/nep-11/count`
    - `[GET]` `/api/v1/tokens/nep-11/{hash}/balanceof/{address}`
  - NEP-17
    - `[GET]` `/api/v1/tokens/nep-17?page={number}&size={number}`
    - `[GET]` `/api/v1/tokens/nep-17/count`
    - `[GET]` `/api/v1/tokens/nep-17/{hash}/balanceof/{address}`
- Contracts
  - `[GET]` `/api/v1/contracts?page={number}&size={number}`
  - `[GET]` `/api/v1/contracts/count`
  - `[GET]` `/api/v1/contracts/{hash}`
  - `[GET]` `/api/v1/contracts/{hash}/abi`
  - `[GET]` `/api/v1/contracts/{hash}/manifest`
  - `[GET]` `/api/v1/contracts/{hash}/nef`
  - `[GET]` `/api/v1/contracts/{hash}/storage`
- Wallet
  - `[POST]` `/api/v1/wallet/open`
  - `[POST]` `/api/v1/wallet/create`
  - `[POST]` `/api/v1/wallet/{session}/address/create`
  - `[GET]` `/api/v1/wallet/{session}/address/list`
  - `[GET]` `/api/v1/wallet/{session}/asset/list`
  - `[GET]` `/api/v1/wallet/{session}/balance/list`
  - `[POST]` `/api/v1/wallet/{session}/changepassword`
  - `[GET]` `/api/v1/wallet/{session}/close`
  - `[GET]` `/api/v1/wallet/{session}/delete/{address}`
  - `[GET]` `/api/v1/wallet/{session}/export/{address}`
  - `[GET]` `/api/v1/wallet/{session}/export`
  - `[GET]` `/api/v1/wallet/{session}/gas/unclaimed`
  - `[GET]` `/api/v1/wallet/{session}/key/list`
  - `[POST]` `/api/v1/wallet/{session}/import`
  - `[POST]` `/api/v1/wallet/{session}/import/multisigaddress`
  - `[POST]` `/api/v1/wallet/{session}/transfer`

Liaojinghui commented:
@superboyiii  can you please review and test it？

superboyiii commented:
This is awesome. Need some time for test.

superboyiii commented:
Build well on my local PC, not sure if it referred to the changes on *csproj of Oracle and Applicationlog
![image](https://github.com/neo-project/neo-modules/assets/38493437/7bc236cb-06b0-414c-9d82-6d332f459c41)

superboyiii commented:
Tried revert the modification from ApplicationLogs and OracleService, forked it to my repo and run CodeQL check. Still failed.

superboyiii commented on src/ApplicationLogs/ApplicationLogs.csproj:
```diff
@@ -5,8 +5,8 @@
   </PropertyGroup>
 
-  <ItemGroup>
-    <ProjectReference Include="..\RpcServer\RpcServer.csproj" AdditionalProperties="IncludeSettingsFileOutput=False"/>
```
Anyway, we have to revert this. Otherwise config.json for applicationlog will have conflict with the one for RpcServer since they will build in the same path.

superboyiii commented on src/OracleService/OracleService.csproj:
```diff
@@ -1,16 +1,16 @@
-  </ItemGroup>
-
-  <ItemGroup>
-    <ProjectReference Include="..\RpcServer\RpcServer.csproj" AdditionalProperties="IncludeSettingsFileOutput=False"/>
```
Please revert. The same reason as above.

cschuchardt88 commented:
In the `code.yml` it says if there is an error `autobuilding` the project to comment out that step and add your own build commands or script to run. For example we needed to do that for `neo-express` repo you have a look at that if you want https://github.com/neo-project/neo-express/blob/master/.github/workflows/code.yml

superboyiii commented on src/RestServer/RestServer.csproj:
```diff
@@ -0,0 +1,24 @@
+    <PackageReference Include="Microsoft.AspNetCore.Mvc.NewtonsoftJson" Version="7.0.11" />
+    <PackageReference Include="Microsoft.AspNetCore.Mvc.Versioning" Version="5.1.0" />
+    <PackageReference Include="Microsoft.AspNetCore.Mvc.Versioning.ApiExplorer" Version="5.1.0" />
+    <PackageReference Include="Neo" Version="3.6.0" />
```
```suggestion
```
Already included in `Directory.Build.props`

superboyiii commented on docs/RestServer/README.md:
```diff
@@ -0,0 +1,136 @@
+- Utils
+  - `[GET]` `/api/v1/utils/{hash}/address`
+  - `[GET]` `/api/v1/utils/{address}/scripthash`
+  - `[GET]` `/api/v1/utils/{hash}/{address}/validate`
```
```suggestion
  - `[GET]` `/api/v1/utils/{hash} or {address}/validate`

```

superboyiii commented on docs/RestServer/README.md:
```diff
@@ -0,0 +1,136 @@
+  - `[GET]` `/api/v1/utils/{address}/scripthash`
+  - `[GET]` `/api/v1/utils/{hash}/{address}/validate`
+- Node
+  - `[GET]` `/api/v1/node`
```
```suggestion

```

superboyiii commented on src/RestServer/Models/Blockchain/AccountDetails.cs:
```diff
@@ -0,0 +1,33 @@
+        /// </summary>
+        /// <example>NNLi44dJNXtDNSBkofB48aTVYtb1zZrNEs</example>
+        public string Address { get; set; }
+        /// <summary>
+        /// Balance of the account.
+        /// </summary>
+        /// <example>10000000</example>
+        public BigInteger Balance { get; set; }
```
```suggestion
        /// <summary>
        /// Balance of the account.
        /// </summary>
        /// <example>10000000</example>
        public BigInteger Balance { get; set; }
        /// <summary>
        /// Decimals of the token.
        /// </summary>
        /// <example>8</example>
        public BigInteger Decimals { get; set; }
```

superboyiii commented on src/RestServer/Extensions/LedgerContractExtensions.cs:
```diff
@@ -0,0 +1,142 @@
+                    var addressHash = new UInt160(key.ToArray().AsSpan(5));
+                    yield return new AccountDetails()
+                    {
+                        ScriptHash = addressHash,
+                        Address = addressHash.ToAddress(protocolSettings.AddressVersion),
+                        Balance = value.GetInteroperable<AccountState>().Balance,
```
```suggestion
                        ScriptHash = addressHash,
                        Address = addressHash.ToAddress(protocolSettings.AddressVersion),
                        Balance = value.GetInteroperable<AccountState>().Balance,
                        Decimals = NativeContract.GAS.Decimals
```

cschuchardt88 commented on src/RestServer/RestServerUtility.JTokens.cs:
```diff
@@ -0,0 +1,302 @@
+                header.PrevHash,
+                header.NextConsensus,
+                Witness = WitnessToJToken(header.Witness, serializer),
+                header.Size,
```
That is for `blockheader` **only**. I don't return any transactions. The reason for this follows, Saves network bandwidth if the requestor doesn't want transactions. So we only return block information for the current header. But I do see what your saying. How the _true_ block size with transactions is different compared to`blockheader`. It can be changed if you want.

superboyiii replied:
Then maybe it's better to rename it as `GetBlockHeaders` instead of `GetBlocks`.

superboyiii replied:
I see, just confused with some request name and result, will make suggestion below.

superboyiii commented on src/RestServer/Controllers/v1/LedgerController.cs:
```diff
@@ -0,0 +1,389 @@
+            return Ok(lstOfBlocks);
+        }
+
+        /// <summary>
+        /// Gets the current block height of the connected node.
+        /// </summary>
+        /// <returns>Full Block Object.</returns>
+        /// <response code="200">Successful</response>
+        /// <response code="400">An error occurred. See Response for details.</response>
+        [HttpGet("blocks/height", Name = "GetBlockHeight")]
+        [ProducesResponseType(StatusCodes.Status200OK, Type = typeof(Block))]
+        public IActionResult GetCurrentBlock()
+        {
+            var currentIndex = NativeContract.Ledger.CurrentIndex(_neosystem.StoreView);
+            var block = NativeContract.Ledger.GetHeader(_neosystem.StoreView, currentIndex);
+            return Ok(block);
+        }
```
```suggestion
        [HttpGet("blockheader/current", Name = "GetCurrnetBlockHeader")]
        [ProducesResponseType(StatusCodes.Status200OK, Type = typeof(Block))]
        public IActionResult GetCurrentBlockHeader()
        {
            var currentIndex = NativeContract.Ledger.CurrentIndex(_neosystem.StoreView);
            var blockheader = NativeContract.Ledger.GetHeader(_neosystem.StoreView, currentIndex);
            return Ok(blockheader);
        }
```

superboyiii replied:
```suggestion
        /// <summary>
        /// Gets the current block header of the connected node.
        /// </summary>
        /// <returns>Full Block Header Object.</returns>
        /// <response code="200">Successful</response>
        /// <response code="400">An error occurred. See Response for details.</response>
        [HttpGet("blockheader/current", Name = "GetCurrnetBlockHeader")]
        [ProducesResponseType(StatusCodes.Status200OK, Type = typeof(Header))]
        public IActionResult GetCurrentBlockHeader()
        {
            var currentIndex = NativeContract.Ledger.CurrentIndex(_neosystem.StoreView);
            var blockheader = NativeContract.Ledger.GetHeader(_neosystem.StoreView, currentIndex);
            return Ok(blockheader);
        }
```

superboyiii commented on src/RestServer/Controllers/v1/LedgerController.cs:
```diff
@@ -0,0 +1,389 @@
+            [FromQuery(Name = "page")]
+            int skip = 1,
+            [FromQuery(Name = "size")]
+            int take = 1)
```
```suggestion
            int take = 50)
```
I'm not sure, but how about make the default count the same as default MaxPageSize or maybe a value between 1 and 50. Even I have only 2 transactions, I still need input page, size into request, seems not convenient.

cschuchardt88 replied:
I did that so if they make the request without `page` or `size` it only gets one. So web crawlers don't eat bandwidth. But sure we can change them all.

superboyiii commented on src/RestServer/Controllers/v1/LedgerController.cs:
```diff
@@ -0,0 +1,389 @@
+            [FromQuery(Name = "page")]
+            int skip = 1,
+            [FromQuery(Name = "size")]
+            int take = 1)
```
```suggestion
            int take = 50)
```

superboyiii commented on src/RestServer/Controllers/v1/LedgerController.cs:
```diff
@@ -0,0 +1,389 @@
+            [FromQuery(Name = "page")]
+            int skip = 1,
+            [FromQuery(Name = "size")]
+            int take = 1)
```
```suggestion
            int take = 50)
```

superboyiii commented on src/RestServer/Controllers/v1/LedgerController.cs:
```diff
@@ -0,0 +1,389 @@
+            [FromQuery(Name = "page")]
+            int skip = 1,
+            [FromQuery(Name = "size")]
+            int take = 1)
```
```suggestion
            int take = 50)
```

superboyiii commented on src/RestServer/Controllers/v1/TokensController.cs:
```diff
@@ -0,0 +1,304 @@
+            [FromQuery(Name = "page")]
+            int skip = 1,
+            [FromQuery(Name = "size")]
+            int take = 1)
```
```suggestion
            int take = 50)
```

cschuchardt88 commented:
Changed `hardfork` names. Now removed prefix.
![image](https://github.com/neo-project/neo-modules/assets/8141309/e6fae569-45d5-48fc-86f1-69079e1f664c)

superboyiii commented on src/RestServer/Controllers/v1/LedgerController.cs:
```diff
@@ -0,0 +1,389 @@
+            if (NativeContract.Ledger.ContainsTransaction(_neosystem.StoreView, hash) == false)
+                throw new TransactionNotFoundException(hash);
+            var tx = NativeContract.Ledger.GetTransaction(_neosystem.StoreView, hash);
+            return Ok(tx.Attributes);
```
Need show more critical details such as `txhash` for conflict attributes, `id`, `response` for oracle attributes.
![1702280467312](https://github.com/neo-project/neo-modules/assets/38493437/b8343a34-d366-473c-8656-4ff3f1ed3a94)

cschuchardt88 commented:
Fixed
![image](https://github.com/neo-project/neo-modules/assets/8141309/11492e03-fb10-43b2-ab38-ac7cad2be9aa)

superboyiii commented on src/RestServer/RestServerUtility.JTokens.cs:
```diff
@@ -0,0 +1,330 @@
+                manifest.SupportedStandards,
+                Extra = manifest.Extra.Count > 0 ?
+                    new JObject(manifest.Extra.Properties.Select(s => new JProperty(s.Key.ToString(), s.Value.AsString()))) :
+                    null,
```
NullReferenceException will be threw here when getting native contract manifest

cschuchardt88 replied:
Which line?

cschuchardt88 commented:
@superboyiii  This may help you 
[Neo API version 1.0.postman_collection.json](https://github.com/neo-project/neo-modules/files/13634762/Neo.API.version.1.0.postman_collection.json)

superboyiii commented:
@cschuchardt88 I can't get any response from `Wallet` model, even can't debug into it. Other models work well. Is it working on yours?

cschuchardt88 commented:
Check the config? Its disabled by default.

superboyiii commented:
> Check the config? Its disabled by default.

Oh, OK. Works now, thanks.

superboyiii commented on src/RestServer/Controllers/v1/WalletController.cs:
```diff
@@ -0,0 +1,646 @@
+            [FromBody]
+            WalletOpenModel model)
+        {
+            if (new FileInfo(model.Path).DirectoryName.StartsWith(AppContext.BaseDirectory, StringComparison.InvariantCultureIgnoreCase) == false)
+                throw new UnauthorizedAccessException(model.Path);
```
Why the wallet has to be in subdirectory of CLI root path? It can be anywhere.

cschuchardt88 replied:
https://github.com/neo-project/neo-modules/pull/810#discussion_r1301965681

cschuchardt88 replied:
I prefer it to be remove, but as you see in comment. Others want it removed and so does `code` analysis.

superboyiii replied:
OK, I see. Let's keep it.

superboyiii replied:
Can we at least make the path range in CLI's root path? `AppContext.BaseDirectory` is the subdirectory path.
![1703668908928](https://github.com/neo-project/neo-modules/assets/38493437/3ab6e966-ad33-4881-ba9f-66b8e343e8b0)

cschuchardt88 replied:
base path is where ever you run `neo-cli.exe` at.

superboyiii replied:
> base path is where ever you run `neo-cli.exe` at.

The `new FileInfo(model.Path).DirectoryName` path doesn't have `\\` in the end, so if I'd like to open a wallet which start with `AppContext.BaseDirectory`, it need be included in a subdirectory of Neo-CLI. Can you see the difference?
![1704248164106](https://github.com/neo-project/neo-modules/assets/38493437/f2cf8f43-29a5-4b1d-b62a-3c5f034cf2e4)
![1704248188652](https://github.com/neo-project/neo-modules/assets/38493437/e699b4fe-c965-4492-a2c7-f6d8aff11a97)
![1704248268920](https://github.com/neo-project/neo-modules/assets/38493437/4f3fdf66-4eb2-4218-803e-03e24f130be1)

cschuchardt88 replied:
Yes i can fix that.

cschuchardt88 replied:
Fixed Problem. Can you test again?
![image](https://github.com/neo-project/neo-modules/assets/8141309/98dff455-a32e-4bfd-aca6-ba61c2b6828d)

superboyiii commented on src/RestServer/Controllers/v1/WalletController.cs:
```diff
@@ -0,0 +1,646 @@
+            var session = WalletSessions[sessionId];
+            session.ResetExpiration();
+            var wallet = session.Wallet;
+            var account = model.PrivateKey == null || model.PrivateKey.Length == 0 ?
+                wallet.CreateAccount() :
+                wallet.CreateAccount(model.PrivateKey);
```
```suggestion
            var account = Wallet.GetPrivateKeyFromWIF(model.Wif) == null || model.Wif.Length == 0 ?
                wallet.CreateAccount() :
                wallet.CreateAccount(Wallet.GetPrivateKeyFromWIF(model.Wif));
```
It's not easy for a user to know his byte[] privatekey, we're all using Wif format privatekey now in Dapp wallets.

superboyiii commented on src/RestServer/Models/Wallet/WalletCreateAccountModel.cs:
```diff
@@ -0,0 +1,24 @@
+    /// <summary>
+    /// Create account object.
+    /// </summary>
+    public class WalletCreateAccountModel
+    {
+        /// <summary>
+        /// Private key of the address you want to create. Can be null or empty.
+        /// </summary>
+        /// <example>CHeABTw3Q5SkjWharPAhgE+p+rGVN9FhlO4hXoJZQqA=</example>
+        public byte[] PrivateKey { get; set; }
+    }
```
```suggestion
    public class WalletCreateAccountModel
    {
        /// <summary>
        /// Wif format Private key of the address you want to create. Can be null or empty.
        /// </summary>
        /// <example>L4EtTnAypiur9UahCn43fpAexTM2ww4BhsPghqvc77rec6bfYwHc</example>
        public string Wif { get; set; }
    }
```

superboyiii commented on src/RestServer/Controllers/v1/WalletController.cs:
```diff
@@ -0,0 +1,646 @@
+            {
+                if (wallet is NEP6Wallet wallet6)
+                    wallet6.Save();
+                return Ok();
```
```suggestion
                return Ok("Address is deleted.");
```
Should return something I think, otherwise I don't know if my request is correct...

cschuchardt88 replied:
Its Rest, so it just returns json models and http status. Just something we need to document.

cschuchardt88 replied:
because if we do that then the response `content-type` won't be correct.

superboyiii replied:
I see

superboyiii commented on src/RestServer/Controllers/v1/WalletController.cs:
```diff
@@ -0,0 +1,646 @@
+                throw new WalletException("Failed to change password!");
+            if (wallet is NEP6Wallet nep6)
+                nep6.Save();
+            return Ok();
```
```suggestion
            return Ok("Password is changed.");
```
Make sure password is changed.

cschuchardt88 replied:
i think http status is good indicator, because it's a `Rest` service. Normally developers use http status for this. Have a look at the top of function code comments. You will see the comments.

superboyiii commented on src/RestServer/Controllers/v1/WalletController.cs:
```diff
@@ -0,0 +1,646 @@
+            var session = WalletSessions[sessionId];
+            session.ResetExpiration();
+            var wallet = session.Wallet;
+            var account = Wallet.GetPrivateKeyFromWIF(model.Wif) == null || model.Wif.Length == 0 ?
+                wallet.CreateAccount() :
+                wallet.CreateAccount(Wallet.GetPrivateKeyFromWIF(model.Wif));
```
```suggestion
            WalletAccount account;
            if (model.Wif == null || model.Wif.Length == 0)
            {
                account = wallet.CreateAccount();
            }
            else
            {
                account = wallet.Import(model.Wif);
            }
```
Sorry, `Wallet.GetPrivateKeyFromWIF()` can't import a null parameter, let's change it like this.

cschuchardt88 replied:
Its Ok for these `Controller` actions to throw exceptions. Server will handle them properly and output as `ErrorModel`. `GetPrivateKeyFromWIF` will throw `ArgumentNullException` or `FormatException`.

But we can change it in this situation. Seeing how the error we will get won't indicate what really is going on. And on top of that we want to `CreateAccount`

superboyiii replied:
> Its Ok for these `Controller` actions to throw exceptions. Server will handle them properly and output as `ErrorModel`. `GetPrivateKeyFromWIF` will throw `ArgumentNullException` or `FormatException`.
I mean if this situation wif is necessary, we can't create a random address using `wallet.CreateAccount();` :)

superboyiii commented on src/RestServer/Controllers/v1/WalletController.cs:
```diff
@@ -0,0 +1,646 @@
+                {
+                    Address = account.Address,
+                    ScriptHash = account.ScriptHash,
+                    Publickey = account.GetKey().PublicKey,
```
```suggestion
                    Publickey = account.GetKey()?.PublicKey,
```
Be care for `multisig` address and `watchonly` address.

superboyiii commented on src/RestServer/Controllers/v1/WalletController.cs:
```diff
@@ -0,0 +1,646 @@
+            session.ResetExpiration();
+            var wallet = session.Wallet;
+            var accounts = new List<WalletAddressModel>();
+            foreach (var account in wallet.GetAccounts())
+                accounts.Add(new WalletAddressModel()
+                {
+                    Address = account.Address,
+                    ScriptHash = account.ScriptHash,
+                    Publickey = account.GetKey().PublicKey,
+                    HasKey = account.HasKey,
+                    Label = account.Label,
+                    WatchOnly = account.WatchOnly,
+                });
```
```suggestion
            using var snapshot = _neosystem.GetSnapshot();
            foreach (var account in wallet.GetAccounts())
            {
                var contract = account.Contract;
                var type = "Nonstandard";

                if (account.WatchOnly)
                {
                    type = "WatchOnly";
                }
                else if (IsMultiSigContract(contract.Script))
                {
                    type = "MultiSignature";
                }
                else if (IsSignatureContract(contract.Script))
                {
                    type = "Standard";
                }
                else if (NativeContract.ContractManagement.GetContract(snapshot, account.ScriptHash) != null)
                {
                    type = "Deployed-Nonstandard";
                }
                accounts.Add(new WalletAddressModel()
                {
                    Address = account.Address,
                    ScriptHash = account.ScriptHash,
                    Publickey = account.GetKey()?.PublicKey,
                    HasKey = account.HasKey,
                    Type = type,
                    Label = account.Label,
                    WatchOnly = account.WatchOnly,
                });
            }
```
Add  address `type`, return null when publickey can't be got when `WatchOnly` or `Multisig`.

cschuchardt88 replied:
You didn't Add `Type` to the Model?

superboyiii commented on src/RestServer/Controllers/v1/WalletController.cs:
```diff
@@ -0,0 +1,646 @@
+using Neo.Wallets;
+using Neo.Wallets.NEP6;
+using System.Net.Mime;
+using System.Numerics;
```
```suggestion
using System.Numerics;
using static Neo.SmartContract.Helper;
```

superboyiii commented on src/RestServer/Models/Wallet/WalletAddressModel.cs:
```diff
@@ -0,0 +1,51 @@
+        /// has a private key or not.
+        /// </summary>
+        /// <example>true</example>
+        public bool HasKey { get; set; }
```
```suggestion
                public bool HasKey { get; set; }
        /// <summary>
        /// Address type.
        /// </summary>
        /// <example>Standard/MultiSignature/WatchOnly</example>
        public string Type { get; set; }
```

superboyiii commented on src/RestServer/Controllers/v1/WalletController.cs:
```diff
@@ -0,0 +1,646 @@
+                {
+                    Address = account.Address,
+                    ScriptHash = account.ScriptHash,
+                    PublicKey = account.GetKey().PublicKey,
```
```suggestion
                    PublicKey = account.GetKey()?.PublicKey,
```

cschuchardt88 commented:
@superboyiii I added you to my repos, so you commit changes directly.

superboyiii commented on src/RestServer/RestServerUtility.JTokens.cs:
```diff
@@ -0,0 +1,330 @@
+        public static JToken SignerToJToken(Signer signer, global::Newtonsoft.Json.JsonSerializer serializer) =>
+            JToken.FromObject(new
+            {
+                Rules = signer.Rules.Select(s => WitnessRuleToJToken(s, serializer)),
```
```suggestion
                Rules = signer.Rules?.Select(s => WitnessRuleToJToken(s, serializer)),
```
In `WalletTransferAssets` for normal NEO or GAS transactions, no `Rules` here.

cschuchardt88 replied:
can you do
```
Rules = signer.Rules != null signer.Rules.Select(s => WitnessRuleToJToken(s, serializer)) : Array.Empty<WitnessRule>(),
```

So that way, it returns an empty array on null. So its consistent with the rest of the json in `RestServer`.

superboyiii commented:
Is this OK? @cschuchardt88 
![1703732686186](https://github.com/neo-project/neo-modules/assets/38493437/bd3924bf-3b8c-411c-b963-d6d643137247)

cschuchardt88 commented:
Has to be this:
```csharp
Rules = signer.Rules != null ? signer.Rules.Select(s => WitnessRuleToJToken(s, serializer)) : Array.Empty<WitnessRule>(),
```

You have to call `WitnessRuleToJToken` or else it can't serialize the object, and you will get an error.

cschuchardt88 commented:
Also i forgot to mention there is console commands.

![image](https://github.com/neo-project/neo-modules/assets/8141309/89c65fef-c404-40a6-ad3e-ccbf232c40c5)

superboyiii reviewed (dismissed):
All good now. If can improve https://github.com/neo-project/neo-modules/pull/839#discussion_r1436873391, will be perfect.

superboyiii commented:
@Jim8y @shargon Need your review for double check.

cschuchardt88 commented:
@superboyiii Did you check out the Swagger UI?

- http://127.0.0.1:10339/swagger/index.html

![image](https://github.com/neo-project/neo-modules/assets/8141309/bcd548f8-4e09-4a03-b041-9e53d570e05f)

superboyiii commented:
> @superboyiii Did you check out the Swagger UI?

Tried it, it's really cool and convinent!

vncoelho reviewed (commented):
remainder to check https://github.com/neo-project/neo-modules/pull/859, remove websocker

shargon commented:
Is not forgotten, I will need time to review it, I really love this PR

vncoelho commented:
I was missing the .xml file in the folder. I will try again.

vncoelho commented:
![image](https://github.com/neo-project/neo-modules/assets/17395762/8d476cb8-164b-4217-b8d4-dcafed76dddf)

vncoelho commented:
```cs
neo> FATAL [13:05:36.078] 	System.IO.FileNotFoundException
Could not load file or assembly 'Microsoft.AspNetCore.Mvc.Versioning.ApiExplorer, Version=5.1.0.0,  Culture=neutral,  PublicKeyToken=adb9793829...ch()
   at  System.Thr...eadStart()
Unhandled exception. System.IO.FileNotFoundException: Could not load file or assembly 'Microsoft.AspNetCore.Mvc.Versioning.ApiExplorer, Version=5.1.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60'. The system cannot find the file specified.

File name: 'Microsoft.AspNetCore.Mvc.Versioning.ApiExplorer, Version=5.1.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60'
   at Neo.Plugins.RestServer.RestWebServer.<Start>b__7_1(IServiceCollection services)
   at Microsoft.AspNetCore.Hosting.WebHostBuilder.BuildCommonServices(AggregateException& hostingStartupErrors)
   at Microsoft.AspNetCore.Hosting.WebHostBuilder.Build()
   at Neo.Plugins.RestServer.RestWebServer.Start()
   at Neo.Plugins.RestServer.RestServerPlugin.OnSystemLoaded(NeoSystem system)
   at Neo.NeoSystem..ctor(ProtocolSettings settings, String storageEngine, String storagePath) in /opt/neoLib/src/Neo/NeoSystem.cs:line 130
   at Neo.CLI.MainService.Start(String[] args) in /opt/neoLib/src/Neo.CLI/CLI/MainService.cs:line 379
   at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__128_1(Object state)
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
   at System.Threading.PortableThreadPool.WorkerThread.WorkerThreadStart()
/opt/start_node.sh: line 2:   111 Aborted                 (core dumped) dotnet neo-cli.dll
root@a262d64d5120:/# 
```

vncoelho commented:
I tried now copying all .dlls to the folder. But still some errors

![image](https://github.com/neo-project/neo-modules/assets/17395762/f3b51438-372a-4ffc-bc24-8db6af1b6f5c)
```cs
NEO-CLI v3.6.2  -  NEO v3.6.2  -  NEO-VM v3.6.2

neo> FATAL [15:12:58.316] 	System.Reflection.ReflectionTypeLoadException
Unable to load one or more of the requested types.
Could not load file or assembly 'Microsoft.EntityFrameworkCore, Version=7.0.10.0,  Culture=neutral,  PublicKeyToken=adb9793829...ch()
   at  System.Thr...eadStart()
Unhandled exception. System.Reflection.ReflectionTypeLoadException: Unable to load one or more of the requested types.
Could not load file or assembly 'Microsoft.EntityFrameworkCore, Version=7.0.10.0, Culture=neutral, PublicKeyToken=adb9793829ddae60'. The system cannot find the file specified.

   at System.Reflection.RuntimeModule.GetTypes(RuntimeModule module)
   at Microsoft.AspNetCore.Mvc.Controllers.ControllerFeatureProvider.PopulateFeature(IEnumerable`1 parts, ControllerFeature feature)
   at Microsoft.AspNetCore.Mvc.ApplicationParts.ApplicationPartManager.PopulateFeature[TFeature](TFeature feature)
   at Microsoft.AspNetCore.Mvc.ApplicationModels.ControllerActionDescriptorProvider.GetControllerTypes()
   at Microsoft.AspNetCore.Mvc.ApplicationModels.ControllerActionDescriptorProvider.GetDescriptors()
   at Microsoft.AspNetCore.Mvc.ApplicationModels.ControllerActionDescriptorProvider.OnProvidersExecuting(ActionDescriptorProviderContext context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.DefaultActionDescriptorCollectionProvider.UpdateCollection()
   at Microsoft.AspNetCore.Mvc.Infrastructure.DefaultActionDescriptorCollectionProvider.Initialize()
   at Microsoft.AspNetCore.Mvc.Infrastructure.DefaultActionDescriptorCollectionProvider.get_ActionDescriptors()
   at Microsoft.AspNetCore.Mvc.ApiExplorer.DefaultApiVersionDescriptionProvider.EnumerateApiVersions(IActionDescriptorCollectionProvider actionDescriptorCollectionProvider)
   at Microsoft.AspNetCore.Mvc.ApiExplorer.DefaultApiVersionDescriptionProvider.LazyApiVersionDescriptions.EnumerateApiVersions()
   at System.Lazy`1.ViaFactory(LazyThreadSafetyMode mode)
   at System.Lazy`1.ExecutionAndPublication(LazyHelper executionAndPublication, Boolean useDefaultConstructor)
   at System.Lazy`1.CreateValue()
   at Microsoft.AspNetCore.Mvc.ApiExplorer.DefaultApiVersionDescriptionProvider.get_ApiVersionDescriptions()
   at Neo.Plugins.RestServer.RestWebServer.<>c__DisplayClass7_3.<Start>b__37(SwaggerUIOptions options)
   at Microsoft.AspNetCore.Builder.SwaggerUIBuilderExtensions.UseSwaggerUI(IApplicationBuilder app, Action`1 setupAction)
   at Neo.Plugins.RestServer.RestWebServer.<Start>b__7_2(IApplicationBuilder app)
   at Microsoft.Extensions.DependencyInjection.AutoRegisterMiddleware.<>c__DisplayClass4_0.<Configure>b__0(IApplicationBuilder app)
   at Microsoft.AspNetCore.Hosting.WebHost.BuildApplication()
   at Microsoft.AspNetCore.Hosting.WebHost.StartAsync(CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Hosting.WebHost.Start()
   at Neo.Plugins.RestServer.RestWebServer.Start()
   at Neo.Plugins.RestServer.RestServerPlugin.OnSystemLoaded(NeoSystem system)
   at Neo.NeoSystem..ctor(ProtocolSettings settings, String storageEngine, String storagePath) in /opt/neoLib/src/Neo/NeoSystem.cs:line 130
   at Neo.CLI.MainService.Start(String[] args) in /opt/neoLib/src/Neo.CLI/CLI/MainService.cs:line 379
   at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__128_1(Object state)
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
   at System.Threading.PortableThreadPool.WorkerThread.WorkerThreadStart()
System.IO.FileNotFoundException: Could not load file or assembly 'Microsoft.EntityFrameworkCore, Version=7.0.10.0, Culture=neutral, PublicKeyToken=adb9793829ddae60'. The system cannot find the file specified.

File name: 'Microsoft.EntityFrameworkCore, Version=7.0.10.0, Culture=neutral, PublicKeyToken=adb9793829ddae60'
/opt/start_node.sh: line 2:    99 Aborted                 (core dumped) dotnet neo-cli.dll
root@2b23b4535fb1:/# 
```

cschuchardt88 commented:
@vncoelho Looks like you are using "Microsoft.EntityFrameworkCore"? You site is programmed in `asp.net mvc`?

vncoelho commented:
Everything is built with SDK `mcr.microsoft.com/dotnet/sdk:7.0.404-jammy`

Node is created with `dotnet publish -c Release -f net7.0`
Plugins are generated with `dotnet publish -c Release -o app`

Client is run with a container with image `mcr.microsoft.com/dotnet/aspnet:7.0.14-jammy`
We used to use aspnet because it was enough for neo-cli to run and simpler and more compact. 
Perhaps we may need another image for client now.

cschuchardt88 commented:
My thoughts are, that you are using the wrong image. Why are you using `web` image like `aspnet:7.0.14-jammy`? Since your doing that way, `Microsoft.Extensions.DependencyInjection.AutoRegisterMiddleware` will; you guessed it auto register. In this case its trying to auto register `Microsoft.EntityFrameworkCore` or something that needs it; but not likely. however i don't use `Microsoft.EntityFrameworkCore` thats for interacting with `SQL` databases.

Maybe also try not add the dependencies that `RestServer` Require. There is an updated list:
![image](https://github.com/neo-project/neo-modules/assets/8141309/84282050-a988-41ea-abcd-9beb9c860938)

vncoelho commented:
I tried just with that dependencies already. I sent the error above.

Maybe I could try with another image.

cschuchardt88 commented:
Different image for sure, But also try disabling swagger in the `config.json`.

vncoelho commented:
Which one in this case?
This involves a change in the packages needed for the client. 

The commands I use are all dockerized.
Any change in this images means some compatibility and additional packages in the images, which leads to more possible leaks for official nodes to use.

If used just by isolated seed nodes without assets management is one thing.

In my vision, simplicity makes security. 
Lets just exactly understand why this need to be changed.

cschuchardt88 commented:
Have you tried docker image `ghcr.io/axlabs/neo3-privatenet-docker/neo-cli-with-plugins:neo-node-3.6.2`?

cschuchardt88 commented:
Also with #3018 you dont need dotnet installed or at all. Another thing is with #3039 I forget to remove `asp.net core` framework libray. from the `csproj`

vncoelho commented:
@cschuchardt88, everything is compiling with the image we current have, except this plugin.
I will try with dotnet sdk image.

vncoelho commented:
Tried with `FROM mcr.microsoft.com/dotnet/sdk:7.0.404-jammy`

```cs
neo> FATAL [11:44:19.675] 	System.Reflection.ReflectionTypeLoadException
Unable to load one or more of the requested types.
Could not load file or assembly 'Microsoft.EntityFrameworkCore, Version=7.0.10.0,  Culture=neutral,  PublicKeyToken=adb9793829...ch()
   at  System.Thr...eadStart()
Unhandled exception. System.Reflection.ReflectionTypeLoadException: Unable to load one or more of the requested types.
Could not load file or assembly 'Microsoft.EntityFrameworkCore, Version=7.0.10.0, Culture=neutral, PublicKeyToken=adb9793829ddae60'. The system cannot find the file specified.

   at System.Reflection.RuntimeModule.GetTypes(RuntimeModule module)
   at Microsoft.AspNetCore.Mvc.Controllers.ControllerFeatureProvider.PopulateFeature(IEnumerable`1 parts, ControllerFeature feature)
   at Microsoft.AspNetCore.Mvc.ApplicationParts.ApplicationPartManager.PopulateFeature[TFeature](TFeature feature)
   at Microsoft.AspNetCore.Mvc.ApplicationModels.ControllerActionDescriptorProvider.GetControllerTypes()
   at Microsoft.AspNetCore.Mvc.ApplicationModels.ControllerActionDescriptorProvider.GetDescriptors()
   at Microsoft.AspNetCore.Mvc.ApplicationModels.ControllerActionDescriptorProvider.OnProvidersExecuting(ActionDescriptorProviderContext context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.DefaultActionDescriptorCollectionProvider.UpdateCollection()
   at Microsoft.AspNetCore.Mvc.Infrastructure.DefaultActionDescriptorCollectionProvider.Initialize()
   at Microsoft.AspNetCore.Mvc.Infrastructure.DefaultActionDescriptorCollectionProvider.get_ActionDescriptors()
   at Microsoft.AspNetCore.Mvc.ApiExplorer.DefaultApiVersionDescriptionProvider.EnumerateApiVersions(IActionDescriptorCollectionProvider actionDescriptorCollectionProvider)
   at Microsoft.AspNetCore.Mvc.ApiExplorer.DefaultApiVersionDescriptionProvider.LazyApiVersionDescriptions.EnumerateApiVersions()
   at System.Lazy`1.ViaFactory(LazyThreadSafetyMode mode)
   at System.Lazy`1.ExecutionAndPublication(LazyHelper executionAndPublication, Boolean useDefaultConstructor)
   at System.Lazy`1.CreateValue()
   at Microsoft.AspNetCore.Mvc.ApiExplorer.DefaultApiVersionDescriptionProvider.get_ApiVersionDescriptions()
   at Neo.Plugins.RestServer.RestWebServer.<>c__DisplayClass7_3.<Start>b__37(SwaggerUIOptions options)
   at Microsoft.AspNetCore.Builder.SwaggerUIBuilderExtensions.UseSwaggerUI(IApplicationBuilder app, Action`1 setupAction)
   at Neo.Plugins.RestServer.RestWebServer.<Start>b__7_2(IApplicationBuilder app)
   at Microsoft.Extensions.DependencyInjection.AutoRegisterMiddleware.<>c__DisplayClass4_0.<Configure>b__0(IApplicationBuilder app)
   at Microsoft.AspNetCore.Hosting.WebHost.BuildApplication()
   at Microsoft.AspNetCore.Hosting.WebHost.StartAsync(CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Hosting.WebHost.Start()
   at Neo.Plugins.RestServer.RestWebServer.Start()
   at Neo.Plugins.RestServer.RestServerPlugin.OnSystemLoaded(NeoSystem system)
   at Neo.NeoSystem..ctor(ProtocolSettings settings, String storageEngine, String storagePath) in /opt/neoLib/src/Neo/NeoSystem.cs:line 130
   at Neo.CLI.MainService.Start(String[] args) in /opt/neoLib/src/Neo.CLI/CLI/MainService.cs:line 379
   at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__128_1(Object state)
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
   at System.Threading.PortableThreadPool.WorkerThread.WorkerThreadStart()
System.IO.FileNotFoundException: Could not load file or assembly 'Microsoft.EntityFrameworkCore, Version=7.0.10.0, Culture=neutral, PublicKeyToken=adb9793829ddae60'. The system cannot find the file specified.

File name: 'Microsoft.EntityFrameworkCore, Version=7.0.10.0, Culture=neutral, PublicKeyToken=adb9793829ddae60'
/opt/start_node.sh: line 2:   101 Aborted                 (core dumped) dotnet neo-cli.dll
```

vncoelho commented:
> > @superboyiii Did you check out the Swagger UI?
> 
> Tried it, it's really cool and convinent!

@superboyiii, 
* I am being able to build and create the plugins .dlls
* However, they are producing errors when running on neo-cli

It is strange because the same images is running in another nodes and everything is working, just the container with rest server is bugging.

vncoelho commented:
@cschuchardt88, setting the "EnableSwagger" flag to false it was able to run.

cschuchardt88 commented:
Its Microsofts Swagger library, I have no clue, they want to load the Entity Framework. I have found bugs with swagger, and Microsoft never said anything on their github page, when i created issue, It's been about 6 months now and 2 bugs reported and now this. 

@superboyiii I think we remove swagger for now.

vncoelho reviewed (commented):
Maybe update the Readme if some ddls from Swager is not needed.

And perhaps this other one is necessary

```cs
neo> FATAL [23:47:51.245] 	System.IO.FileNotFoundException
Could not load file or assembly 'Microsoft.AspNetCore.Mvc.Versioning.ApiExplorer, Version=5.1.0.0,  Culture=neutral,  PublicKeyToken=adb9793829...ch()
   at  System.Thr...eadStart()
Unhandled exception. System.IO.FileNotFoundException: Could not load file or assembly 'Microsoft.AspNetCore.Mvc.Versioning.ApiExplorer, Version=5.1.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60'. The system cannot find the file specified.

File name: 'Microsoft.AspNetCore.Mvc.Versioning.ApiExplorer, Version=5.1.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60'
   at Neo.Plugins.RestServer.RestWebServer.<Start>b__7_1(IServiceCollection services)
   at Microsoft.AspNetCore.Hosting.WebHostBuilder.BuildCommonServices(AggregateException& hostingStartupErrors)
   at Microsoft.AspNetCore.Hosting.WebHostBuilder.Build()
   at Neo.Plugins.RestServer.RestWebServer.Start()
   at Neo.Plugins.RestServer.RestServerPlugin.OnSystemLoaded(NeoSystem system)
   at Neo.NeoSystem..ctor(ProtocolSettings settings, String storageEngine, String storagePath) in /opt/neoLib/src/Neo/NeoSystem.cs:line 130
   at Neo.CLI.MainService.Start(String[] args) in /opt/neoLib/src/Neo.CLI/CLI/MainService.cs:line 379
   at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__128_1(Object state)
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
   at System.Threading.PortableThreadPool.WorkerThread.WorkerThreadStart()
/opt/start_node.sh: line 2:   161 Aborted                 (core dumped) dotnet neo-cli.dll
```

vncoelho commented:
```bash
if [ $PLUGIN_TO_INCLUDE = "RestServer" ]; then
    echo "Going to copy file $ORIGIN_PATH/$PLUGIN_TO_INCLUDE.dll TO $DEST_FOLDER - and some other dependencies"
    cp -ri $ORIGIN_PATH/$PLUGIN_TO_INCLUDE.dll $DEST_FOLDER
    cp -ri $ORIGIN_PATH/Microsoft.AspNetCore.JsonPatch.dll $DEST_FOLDER
    cp -ri $ORIGIN_PATH/Microsoft.AspNetCore.Mvc.NewtonsoftJson.dll $DEST_FOLDER
    cp -ri $ORIGIN_PATH/Microsoft.OpenApi.dll $DEST_FOLDER
    cp -ri $ORIGIN_PATH/Swashbuckle.AspNetCore.Swagger.dll $DEST_FOLDER
    cp -ri $ORIGIN_PATH/Swashbuckle.AspNetCore.SwaggerGen.dll $DEST_FOLDER
    cp -ri $ORIGIN_PATH/Swashbuckle.AspNetCore.SwaggerUI.dll $DEST_FOLDER
    cp -ri $ORIGIN_PATH/*.dll $DEST_FOLDER    
    cp -ri $ORIGIN_PATH/RestServer.xml $DEST_FOLDER   
fi
```

I did not identified all one by one.
But to work I needed to copy all `    cp -ri $ORIGIN_PATH/*.dll $DEST_FOLDER    `

vncoelho commented:
```html
code | -2146232830
name | "ReflectionTypeLoadException"
message | "Unable to load one or more of the requested types.\nCould not load file or assembly 'Microsoft.EntityFrameworkCore, Version=7.0.10.0, Culture=neutral, PublicKeyToken=adb9793829ddae60'. The system cannot find the file specified.\n"
```

Error when I entered in the main url

vncoelho commented:
@cschuchardt88,

I see that you opened 3 big PRs in this repo.
Maybe it is time to migrate this repo to Neo-Core as well and you move that PRs in the sequence.
I believe it will be better for us and our workflow.

shargon commented:
> @cschuchardt88,
> 
> I see that you opened 3 big PRs in this repo. Maybe it is time to migrate this repo to Neo-Core as well and you move those PRs in the sequence. I believe it will be better for us and our workflow.

After we merge the big PR we can do that

cschuchardt88 commented:
> ```js
> code | -2146232830
> name | "ReflectionTypeLoadException"
> message | "Unable to load one or more of the requested types.\nCould not load file or assembly 'Microsoft.EntityFrameworkCore, Version=7.0.10.0, Culture=neutral, PublicKeyToken=adb9793829ddae60'. The system cannot find the file specified.\n"
> ```
> 
> Error when I accessed the main url of the api

Must be the configuration of the docker image. Have you tried full version of dotnet (docker image)? and not a lite version?

shargon commented on src/RestServer/RestServerPlugin.Console.cs:
```diff
@@ -0,0 +1,131 @@
+            }
+        }
+
+        private string ReadUserInput(string prompt, bool password = false)
```
Move this method to `ConsoleHelper`?

cschuchardt88 replied:
It would be better if `Neo.ConsoleService` made it `public`. It's the same one i copied it from there.

shargon commented on src/RestServer/Extensions/LedgerContractExtensions.cs:
```diff
@@ -0,0 +1,144 @@
+        private const byte _prefix_account = 20;
+        private const byte _prefix_totalsupply = 11;
+
+        public static IEnumerable<(StorageKey key, StorageItem value)> GetStorageByPrefix(this ContractState contractState, DataCache snapshot, byte[] prefix)
+        {
+            ArgumentNullException.ThrowIfNull(nameof(contractState));
+            ArgumentNullException.ThrowIfNull(nameof(snapshot));
+            if (prefix?.Length == 0)
+                throw new ArgumentNullException(nameof(prefix));
+            foreach (var (key, value) in snapshot.Find(StorageKey.CreateSearchPrefix(contractState.Id, prefix)))
+                yield return (key, value);
+        }
+
+        public static StorageItem GetStorageByKey(this ContractState contractState, DataCache snapshot, byte[] storageKey)
+        {
+            ArgumentNullException.ThrowIfNull(nameof(contractState));
+            ArgumentNullException.ThrowIfNull(nameof(snapshot));
+            if (storageKey?.Length == 0)
+                throw new ArgumentNullException(nameof(storageKey));
+            foreach (var (key, value) in snapshot.Find(StorageKey.CreateSearchPrefix(contractState.Id, storageKey)))
+                if (key.Key.Span.SequenceEqual(storageKey))
+                    return value;
+            return default;
+        }
```
Some of these methos are not used, it can be removed

cschuchardt88 replied:
I'm not sure. I think `ContractController` uses for contract storage requests.

shargon commented on src/RestServer/Newtonsoft/Json/BigDecimalJsonConverter.cs:
```diff
@@ -0,0 +1,48 @@
+
+        public override void WriteJson(JsonWriter writer, BigDecimal value, JsonSerializer serializer)
+        {
+            var o = JToken.FromObject(new
+            {
+                value.Value,
+                value.Decimals,
+            }, serializer);
+            o.WriteTo(writer);
```
use JObject if is not possible to use float?  (123.1234 instead of {value:123,decimals:1234} )

cschuchardt88 replied:
This uses `Newtonsoft.Json`.

shargon replied:
I think that ethereum use "decimal style" (1.234) for api

shargon commented on src/RestServer/RestServer.csproj:
```diff
@@ -0,0 +1,25 @@
+    <PackageId>Neo.Plugins.RestServer</PackageId>
+    <RootNamespace>Neo.Plugins.RestServer</RootNamespace>
+    <ImplicitUsings>enable</ImplicitUsings>
+    <Nullable>disable</Nullable>
```
```suggestion
    <Nullable>enable</Nullable>
```

cschuchardt88 replied:
We will need to fix all the warning too or errors.

cschuchardt88 commented on src/RestServer/Controllers/v1/NodeController.cs:
```diff
@@ -0,0 +1,86 @@
+        /// <response code="400">An error occurred. See Response for details.</response>
+        [HttpGet("peers", Name = "GetNodeRemotePeers")]
+        [ProducesResponseType(StatusCodes.Status200OK, Type = typeof(RemoteNodeModel[]))]
+        public IActionResult GetPeers()
```
That can be a possibility. I was just following `RpcServer` methods. Also added some additional ones.

shargon replied:
Currently all is protected or not, so is good for me (`RpcServer` style), is just an idea for the future.

shargon commented:
Awesome work @cschuchardt88 👏 for me is ready to merge when nullable is fixed

shargon commented on src/RestServer/Controllers/v1/WalletController.cs:
```diff
@@ -0,0 +1,676 @@
+            [FromBody]
+            WalletOpenModel model)
+        {
+            if (AppContext.BaseDirectory.StartsWith(new FileInfo(model.Path).DirectoryName, StringComparison.InvariantCultureIgnoreCase) == false)
```
`c:\myvalidPath\myValidPath2\..\..\` it will work

cschuchardt88 replied:
Good catch. I'll take your recommendation on how to fix. Also if you want you can update this `method`.

vncoelho commented:
> > ```js
> > code | -2146232830
> > name | "ReflectionTypeLoadException"
> > message | "Unable to load one or more of the requested types.\nCould not load file or assembly 'Microsoft.EntityFrameworkCore, Version=7.0.10.0, Culture=neutral, PublicKeyToken=adb9793829ddae60'. The system cannot find the file specified.\n"
> > ```
> > 
> > Error when I accessed the main url of the api
> 
> Must be the configuration of the docker image. Have you tried full version of dotnet (docker image)? and not a lite version?

https://hub.docker.com/_/microsoft-dotnet

Which one of these images?

shargon commented on src/RestServer/Controllers/v1/WalletController.cs:
```diff
@@ -0,0 +1,676 @@
+            [FromBody]
+            WalletOpenModel model)
+        {
+            if (AppContext.BaseDirectory.StartsWith(new FileInfo(model.Path).DirectoryName, StringComparison.InvariantCultureIgnoreCase) == false)
```
```suggestion
            if (AppContext.BaseDirectory.StartsWith(Path.GetDirectoryName(Path.GetFullPath(model.Path)), StringComparison.InvariantCultureIgnoreCase) == false)
```

cschuchardt88 commented:
@vncoelho ill look into it, its going to take some time to set up.

cschuchardt88 commented:
@vncoelho If you want to get up and running now

### use `jammy`
`docker pull ubuntu` 

### Follow this
https://learn.microsoft.com/en-us/dotnet/core/install/linux-scripted-manual#scripted-install

----------------------------------------
Aslo have you tried
```dockerfile
FROM --platform=$BUILDPLATFORM mcr.microsoft.com/dotnet/sdk:7.0 AS Build

COPY neo-cli /neo-cli
COPY Neo.ConsoleService /Neo.ConsoleService
COPY NuGet.Config /neo-cli

WORKDIR /neo-cli
RUN dotnet restore && dotnet publish -c Release -o /app

FROM --platform=$BUILDPLATFORM mcr.microsoft.com/dotnet/aspnet:7.0 AS Final
RUN apt-get update && apt-get install -y \
  screen \
  libleveldb-dev \
  sqlite3
RUN rm -rf /var/lib/apt/lists/*

WORKDIR /neo-cli
COPY  --from=Build /app .

ENTRYPOINT ["screen","-DmS","node","dotnet","neo-cli.dll","-r"]
```

vncoelho commented:
> @vncoelho If you want to get up and running now
> 
> ### use `jammy`
> `docker pull ubuntu` 
> 
> ### Follow this
> https://learn.microsoft.com/en-us/dotnet/core/install/linux-scripted-manual#scripted-install
> 
> ----------------------------------------
> Aslo have you tried
> ```dockerfile
> FROM --platform=$BUILDPLATFORM mcr.microsoft.com/dotnet/sdk:7.0 AS Build
> 
> COPY neo-cli /neo-cli
> COPY Neo.ConsoleService /Neo.ConsoleService
> COPY NuGet.Config /neo-cli
> 
> WORKDIR /neo-cli
> RUN dotnet restore && dotnet publish -c Release -o /app
> 
> FROM --platform=$BUILDPLATFORM mcr.microsoft.com/dotnet/aspnet:7.0 AS Final
> RUN apt-get update && apt-get install -y \
>   screen \
>   libleveldb-dev \
>   sqlite3
> RUN rm -rf /var/lib/apt/lists/*
> 
> WORKDIR /neo-cli
> COPY  --from=Build /app .
> 
> ENTRYPOINT ["screen","-DmS","node","dotnet","neo-cli.dll","-r"]
> ```

I do not understand your message. 
You say to download Ubuntu.

I only use Ubuntu or other linux distro.


This image you send in the end of the message is a complete sdk.


The point is, state very clear what the changes requests. This change in my main image looks like something that introduces thousands of new packages for a simple seed or consensus node. I believe we have tried to go for simplicity in the past (@erikzhang),even that it is still not very clear for everybody.

vncoelho commented:
I am seeing the dockerfile you send now.
Let me understand, it is both building and running.

vncoelho commented:
> Its two options you can try, Full version of Ubuntu
> 
> under the horizontal line is an other option to try.

That is a drastic change for a PR that simple was intended to remove AspNet. All this should had been discussed and stated before the change. Too much instability.

cschuchardt88 commented:
Its the removal of `asp.net core`, its just my code.

cschuchardt88 commented:
```dockerfile
ARG REPO=mcr.microsoft.com/dotnet/runtime-deps

# Installer image
FROM amd64/buildpack-deps:jammy-curl AS installer

# Retrieve ASP.NET Core
RUN aspnetcore_version=7.0.14 \
    && curl -fSL --output aspnetcore.tar.gz https://dotnetcli.azureedge.net/dotnet/aspnetcore/Runtime/$aspnetcore_version/aspnetcore-runtime-$aspnetcore_version-linux-x64.tar.gz \
    && aspnetcore_sha512='00f55556cb580d81bf0059a61a642ed8b405452d55e94460c03a0edec9a4f608fd78561560e5fc5bf6e42fb1f45420eba75f8d102d8bd46686379dab7ffde6f6' \
    && echo "$aspnetcore_sha512  aspnetcore.tar.gz" | sha512sum -c - \
    && tar -oxzf aspnetcore.tar.gz ./shared/Microsoft.AspNetCore.App \
    && rm aspnetcore.tar.gz


# ASP.NET Core image
FROM $REPO:7.0.14-jammy-amd64

# ASP.NET Core version
ENV ASPNET_VERSION=7.0.14

COPY --from=installer ["/shared/Microsoft.AspNetCore.App", "/usr/share/dotnet/shared/Microsoft.AspNetCore.App"]
```

vncoelho commented:
> ```dockerfile
> mcr.microsoft.com/dotnet/runtime-deps
> ```

That looks like to be too much specific configurations, @cschuchardt88.

From this code I could see that maybe [dotnet/runtime-deps](https://hub.docker.com/_/microsoft-dotnet-runtime-deps/): .NET Runtime Dependencies is better than [dotnet/aspnet](https://hub.docker.com/_/microsoft-dotnet-aspnet/): ASP.NET Core Runtime for this PR. But it is not clear to me.

cschuchardt88 commented:
Above was a modified version of the `asp.net core` `dockerfile`.

vncoelho commented:
> Above was a modified version of the `asp.net core` `dockerfile`.

I do not understand this still, @cschuchardt88.
A modified version of asp.net core...aehauheauea

I think we should do something very very simpler for now. Keeping compatibility at max.
Later (when we merge devpack and module), and when we have a complete integrate environment (more workflows as you are proposing in other PRs), when we optimize and improve packages and these things. At least for me, it is becoming a little bit more complex for what was quite simple.

cschuchardt88 commented:
What you can did is use asp.net core image add Argument `REPO` and set it two `mcr.microsoft.com/dotnet/runtime-deps`

cschuchardt88 commented:
I have ran into this problem before with docker, and my `restserver`

cschuchardt88 commented:
![image](https://github.com/neo-project/neo-modules/assets/8141309/3a7b1bfd-caf1-4afe-a547-116637f2c7ee)

Works fine

vncoelho commented on src/RestServer/config.json:
```diff
@@ -0,0 +1,27 @@
+    "EnableCompression": true,
+    "CompressionLevel": "SmallestSize", // "Fastest" or "Optimal" or "NoCompression" or "SmallestSize"
+    "EnableForwardedHeaders": false,
+    "EnableSwagger": true,
```
@cschuchardt88 ping

cschuchardt88 replied:
what do you mean?

vncoelho replied:
Swagger was removed, no?

cschuchardt88 replied:
Its not removed. if you want `swagger` loaded then set to `true` else `false` and the swagger `dlls` are **not** `required` if `false`

shargon replied:
Please don't remove swagger 🙏

vncoelho commented:
@shargon, the error was `Sqlitewallet` auto dependency injection.
So, I think we are fine now. Everything I tested here is ok.
Are we ready to go?

vncoelho reviewed (commented):
https://node3-neo3-rest.neocompiler.io/swagger/index.html

cschuchardt88 commented:
very nice

shargon commented on src/RestServer/Controllers/v1/WalletController.cs:
```diff
@@ -0,0 +1,674 @@
+                throw new FormatException(nameof(model.Path));
+            if (AppContext.BaseDirectory.StartsWith(path, StringComparison.InvariantCultureIgnoreCase) == false)
+                throw new UnauthorizedAccessException(model.Path);
+            if (System.IO.File.Exists(model.Path) == false)
```
Is checked before

shargon commented on src/RestServer/Controllers/v1/WalletController.cs:
```diff
@@ -0,0 +1,676 @@
+                throw new FormatException(nameof(model.Path));
+            if (AppContext.BaseDirectory.StartsWith(dir, StringComparison.InvariantCultureIgnoreCase) == false)
+                throw new UnauthorizedAccessException(path);
+            if (System.IO.File.Exists(path) == false)
```
@cschuchardt88 maybe we can set a wallet folder in Settings, and just receive the filename?

cschuchardt88 replied:
Trying to make it like `RpcServer` and @superboyiii said the same, they should be able to search anywhere. If we lock down to only one folder than `dapps` or really `exchanges` will not have the ability to securely block an attacker from guessing files or having access to all wallets potentially _(Only through Brute Force).

roman-khimov replied:
Just as a thought --- do we really need `WalletOpen`? How is it used?

To me it's the same question about JSON-RPC `openwallet`, we have never implemented this in NeoGo (see https://github.com/nspcc-dev/neo-go/blob/master/docs/rpc.md#unsupported-methods) and after years of operation no one has ever complained about it. I'm feeling uncomfortable thinking about remote API allowing to open some wallet on the server and then having it widely exposed to everyone around (IIUC it's session-bound here which is much better than what JSON-RPC currently provides, but still).

Yeah, we get some easier APIs to create transactions after that, but even without them we have enough data to create/sign a transaction client-side which is what's happening in every wallet anyway without any issues. We have SDKs in lots of languages that can help doing that. And the key stays client-side, much less exposed and much better controlled. I think this model is safer and given that this plugin is new we may think of not providing this functionality at all, leaving key management out of scope.

cschuchardt88 replied:
The only thing that comes to mind; could be for `exchanges` and `neo-express`. I agree with you @roman-khimov. It should be removed. However its disabled in `config.json` by default. Or even have a separate `plugin` to add in the functionality for `wallets`.

vncoelho replied:
> The only thing that comes to mind; could be for `exchanges` and `neo-express`. I agree with you @roman-khimov. It should be removed. However its disabled in `config.json` by default. Or even have a separate `plugin` to add in the functionality for `wallets`.

Just remove it from here, I think that is enough for now.
In fact, this function has a bad history. If someone wants to use it we add later.

roman-khimov commented on src/RestServer/Controllers/v1/LedgerController.cs:
```diff
@@ -0,0 +1,387 @@
+        [ProducesResponseType(StatusCodes.Status200OK, Type = typeof(AccountDetails[]))]
+        public IActionResult ShowGasAccounts()
+        {
+            var accounts = NativeContract.GAS.ListAccounts(_neosystem.StoreView, _neosystem.Settings);
```
Consider mainnet and some public server providing this API. Won't it be easy to DOS it by doing these `accounts` requests?

cschuchardt88 replied:
It would be something to test. I am using the same logic as `core`; that uses the `storage` `API`. But i do see your point.

AnnaShaleva replied:
We can't have these methods unrestricted.

roman-khimov commented on src/RestServer/Controllers/v1/LedgerController.cs:
```diff
@@ -0,0 +1,387 @@
+        /// <returns>An array of the Transaction object.</returns>
+        /// <response code="200">Successful</response>
+        /// <response code="400">An error occurred. See Response for details.</response>
+        [HttpGet("memorypool", Name = "GetMemoryPoolTransactions")]
```
Why `memorypool` is a part of `ledger`? I'd say it's a part of `node`, there is nothing here that is a part of the ledger yet.

Also, `ledger` can be associated with the respective native contract.

cschuchardt88 replied:
`True`. Good point. I was thinking more that, `node` would be `settings`, `peers`, `connections`. However for `ledger` that would be, more like `transactions` and `blocks`; the actual data related the ledger as in; the stuff that keeps track of the accounts and data.

roman-khimov commented on src/RestServer/Controllers/v1/LedgerController.cs:
```diff
@@ -0,0 +1,387 @@
+        {
+            if (skip < 0 || take < 0 || take > RestServerSettings.Current.MaxPageSize)
+                throw new InvalidParameterRangeException();
+            return Ok(_neosystem.MemPool.Skip((skip - 1) * take).Take(take));
```
Memory pool can change from request to request, I'm not sure paging really fits here.

cschuchardt88 replied:
Yes i thought about that too. But if we were as busy as `etherium` and blocks can have `10,000` transactions in it. I figure paging would be good.

AnnaShaleva replied:
Vote up for paging removal, no practical usage for this option.

roman-khimov commented on src/RestServer/Exceptions/RestErrorCodes.cs:
```diff
@@ -0,0 +1,19 @@
+{
+    internal static class RestErrorCodes
+    {
+        //=========================Rest Codes=========================
```
The API is different, but https://github.com/neo-project/proposals/pull/156 is likely still applicable, just two codes are not really usable.

AnnaShaleva replied:
Right in the proposal. We should have some unification between REST and RPC codes.

cschuchardt88 commented:
@shargon @Jim8y 
This already been tested and reviewed by @superboyiii. However @vncoelho caused a seen, that ended up being false.

shargon commented:
I will review it likely the next week (working of bug fixes now).

shargon commented on src/RestServer/Newtonsoft/Json/BigDecimalJsonConverter.cs:
```diff
@@ -0,0 +1,64 @@
+using Newtonsoft.Json.Linq;
+using System.Numerics;
+
+namespace Neo.Plugins.RestServer.Newtonsoft.Json
```
Cal we move all of these converters to `Neo.Json` project? 

- More organized.
- Reusable.
- Two PR, easy to review.

cschuchardt88 replied:
They use `NewtonSoft.Json` library.

Jim8y replied:
Orderd a new machine, will arrive two days later. My current machine is too weak.

BTW, how hard is it for you to build a Restful Mock testing system?

cschuchardt88 replied:
Well you have to mock the asp.net core. Which can be a pain. After that it's pretty easy.

shargon replied:
Let's move it to `Neo.Json`, easier to approve by chunks

cschuchardt88 replied:
> Let's move it to `Neo.Json`, easier to approve by chunks

Won't be able to do because of **_circular dependencies_**.

cschuchardt88 commented:
Ready for merge https://github.com/neo-project/neo-modules/pull/839#discussion_r1530025463

AnnaShaleva commented on docs/RestServer/Addons.md:
```diff
@@ -0,0 +1,108 @@
+## RestServer Plugin
+In this section of you will learn how to make a `neo-cli` plugin that integrates with `RestServer`
+plugin. Lets take a look at [Example Plugin](/examples/RestServerPlugin).
```
Do we still have an example? Because currently an example plugin link navigates to nowhere.

cschuchardt88 replied:
I think this is a start https://github.com/cschuchardt88/neo-modules/blob/RestServer/docs/RestServer/Addons.md however looks like some stuff is missing.

AnnaShaleva commented on docs/RestServer/Addons.md:
```diff
@@ -0,0 +1,108 @@
+  - **Type** - _Must have a base class of [error](#error-class)._
+
+## Error Class
+Needs to be the same as `RestServer` of else there will be some inconsistencies
```
s/`of else`/`otherwise`

AnnaShaleva commented on docs/RestServer/Addons.md:
````diff
@@ -0,0 +1,108 @@
+## Error Class
+Needs to be the same as `RestServer` of else there will be some inconsistencies
+with end users not knowing which type to use. This class can be `public` or `internal`.
+Properties `Code`, `Name` and `Message` values can be whatever you desire.
+
+**Model**
+```csharp
+public class ErrorModel
+{
+    public int Code { get; set; };
+    public string Name { get; set; };
+    public string Message { get; set; };
+}
````
The case is they can't. We have https://github.com/neo-project/proposals/pull/156 for RPC errors, and if we're merging this PR, then I think we need some unified set of error codes for REST as far.

AnnaShaleva commented on docs/RestServer/Addons.md:
````diff
@@ -0,0 +1,108 @@
+```
+Notice that the _above_ example also returns with HTTP status code of `204 No Content`.
+This action `route` also extends the `contracts` API. Adding method `sayHello`. Routes
+can be what you like as well. But if you want to extend on any existing controller you
````
s/extend on/extend

AnnaShaleva commented on src/RestServer/Binder/UInt160Binder.cs:
```diff
@@ -0,0 +1,47 @@
+
+namespace Neo.Plugins.RestServer.Binder
+{
+    internal class UInt160Binder : IModelBinder
```
How about the same provider for UInt256?

cschuchardt88 replied:
There is no alternative formatting for `UInt256`. Unlike `UInt160` can be `scripthash` or `160HASH` format.

AnnaShaleva commented on src/RestServer/Controllers/v1/ContractsController.cs:
```diff
@@ -0,0 +1,217 @@
+            var contracts = NativeContract.ContractManagement.ListContracts(_neosystem.StoreView);
+            if (contracts.Any() == false)
+                return NoContent();
+            var contractRequestList = contracts.OrderBy(o => o.Manifest.Name).Skip((skip - 1) * take).Take(take);
```
Should be ordered by ID, otherwise paging won't work properly.

cschuchardt88 replied:
I figured `by name` would be more meaningful. But we can change it.

AnnaShaleva commented on src/RestServer/Controllers/v1/ContractsController.cs:
```diff
@@ -0,0 +1,217 @@
+        {
+            if (NativeContract.IsNative(scripthash))
+                return NoContent();
+            var contract = NativeContract.ContractManagement.GetContract(_neosystem.StoreView, scripthash);
+            if (contract == null)
+                throw new ContractNotFoundException(scripthash);
+            var contractStorage = contract.GetStorage(_neosystem.StoreView);
+            return Ok(contractStorage.Select(s => new KeyValuePair<ReadOnlyMemory<byte>, ReadOnlyMemory<byte>>(s.key.Key, s.value.Value)));
```
The server will die on attempt to return the whole storage of some large contract. We need restrictions for this method, paging will suite.

AnnaShaleva commented on src/RestServer/Controllers/v1/ContractsController.cs:
```diff
@@ -0,0 +1,217 @@
+        }
+
+        /// <summary>
+        /// Invoke a method as ReadOnly Flag on a smart contract.
```
Documentation should be adjusted, it doesn't match the behaviour.

AnnaShaleva commented on src/RestServer/Controllers/v1/ContractsController.cs:
```diff
@@ -0,0 +1,217 @@
+        /// <response code="400">An error occurred. See Response for details.</response>
+        [HttpPost("{hash:required}/invoke", Name = "InvokeContractMethod")]
+        [ProducesResponseType(StatusCodes.Status200OK, Type = typeof(ExecutionEngineModel))]
+        public IActionResult InvokeContract(
```
How about the rest of contract-related functionality? E.g. `invokecontractverify` and etc.

AnnaShaleva commented on src/RestServer/Controllers/v1/LedgerController.cs:
```diff
@@ -0,0 +1,388 @@
+        {
+            if (skip < 1 || take < 1 || take > RestServerSettings.Current.MaxPageSize)
+                throw new InvalidParameterRangeException();
+            //var start = (skip - 1) * take + startIndex;
+            //var end = start + take;
```
Remove these comments.

AnnaShaleva commented on src/RestServer/Controllers/v1/NodeController.cs:
```diff
@@ -0,0 +1,86 @@
+        /// <returns>Protocol Settings Object.</returns>
+        /// <response code="200">Successful</response>
+        /// <response code="400">An error occurred. See Response for details.</response>
+        [HttpGet("settings", Name = "GetNodeProtocolSettings")]
+        [ProducesResponseType(StatusCodes.Status200OK, Type = typeof(ProtocolSettingsModel))]
+        public IActionResult GetSettings() =>
+            Ok(_neosystem.Settings.ToModel());
+    }
```
Can be extended to be similar to `getversion` response.

AnnaShaleva commented on src/RestServer/Controllers/v1/TokensController.cs:
```diff
@@ -0,0 +1,303 @@
+            var tokenList = NativeContract.ContractManagement.ListContracts(_neosystem.StoreView);
+            var vaildContracts = tokenList
+                .Where(ContractHelper.IsNep17Supported)
+                .OrderBy(o => o.Manifest.Name)
```
Ditto, sorting by name is a bad idea.

AnnaShaleva commented on src/RestServer/config.json:
```diff
@@ -0,0 +1,24 @@
+    "KeepAliveTimeout": 120,
+    "SslCertFile": "",
+    "SslCertPassword": "",
+    "TrustedAuthorities": [], // Example: "aec13cdd5ea6a3998aec14ac331ad96bedbb770f" (Thumbprint)
+    "EnableBasicAuthentication": false,
+    "RestUser": "",
+    "RestPass": "",
+    "EnableCors": true,
+    "AllowOrigins": [], // Example: "http://www.example.com"
+    "DisableControllers": [],
+    "EnableCompression": true,
+    "CompressionLevel": "SmallestSize", // "Fastest" or "Optimal" or "NoCompression" or "SmallestSize"
```
No comments are allowed by JSON specification. If some comments are needed, then they can be mentioned in the markdown documentation.

Jim8y replied:
> No comments are allowed by JSON specification. If some comments are needed, then they can be mentioned in the markdown documentation.

This is not necessary. Use it in neo core already.

cschuchardt88 replied:
Microsoft `dotnet` allow comments.

shargon commented:
Repository moved to [neo](https://github.com/neo-project/neo), please re-open there
